eyecite/resolve.py

```diff
@@ -0,0 +1,180 @@
+    resolve_shortcase_citation: Callable[
+        [ShortCaseCitation, Dict[ResourceType, FullCaseCitation]],
+        Optional[ResourceType],
+    ] = _resolve_shortcase_citation,
+    resolve_supra_citation: Callable[
+        [SupraCitation, Dict[ResourceType, FullCaseCitation]],
+        Optional[ResourceType],
+    ] = _resolve_supra_citation,
+    resolve_id_citation: Callable[
+        [IdCitation, ResourceType], Optional[ResourceType]
+    ] = _resolve_id_citation,
+) -> Dict[ResourceType, List[CitationBase]]:
+    """Resolves a list of citations to their associated resources by matching
+    each type of Citation object (FullCaseCitation, ShortCaseCitation,
+    SupraCitation, and IdCitation) to a "resource" object. Assumes that the
+    list of citations is ordered in the order that they were extracted from
+    the text (i.e., assumes that supra citations and id citations can only
+    refer to previous references). Returns a dict in the following format:
+        keys = resources
+        values = lists of citations
+
+    The individual resolution steps can be supplanted with more complex logic
+    by passing custom functions (e.g., if you have a thicker resource
+    abstraction that you want to use); the default approach is to use simple
+    heuristics to narrow down the set of possible resolutions. If a citation
+    cannot be definitively resolved to a resource, it is dropped and not
+    resolved.
+    """
+    # Dict of all citation resolutions
+    resolutions: Dict[ResourceType, List[CitationBase]] = defaultdict(list)
+
+    # Dict mapping resources to resolved full citations
+    resolved_full_cites: Dict[ResourceType, FullCaseCitation] = {}
+
+    # The resource of the most recently resolved citation, if any
+    last_resolution: Optional[ResourceType] = None
+
+    # Iterate over each citation and attempt to resolve it to a resource
+    for citation in citations:
+
+        # If the citation is to a non-opinion document, ignore for now
+        if isinstance(citation, NonopinionCitation):
+            pass
+
+        # If the citation is a full case citation, try to resolve it
+        elif isinstance(citation, FullCaseCitation):
+            resolution = resolve_fullcase_citation(citation)
+            resolved_full_cites[resolution] = citation
+
+        # If the citation is a short case citation, try to resolve it
+        elif isinstance(citation, ShortCaseCitation):
+            resolution = resolve_shortcase_citation(
+                citation, resolved_full_cites
+            )
+
+        # If the citation is a supra citation, try to resolve it
+        elif isinstance(citation, SupraCitation):
+            resolution = resolve_supra_citation(citation, resolved_full_cites)
+
+        # If the citation is an id citation, try to resolve it
+        elif isinstance(citation, IdCitation):
+            resolution = resolve_id_citation(citation, last_resolution)
+
+        if resolution:
```

Comment:
Won't this be undefined if the citation is a `NonOpinionCitation`?

Reply:
Yes, good catch.